tiledb/sparse_array.py

```diff
@@ -0,0 +1,455 @@
+
+    def _read_sparse_subarray(self, subarray, attr_names: list, cond, layout):
+        from collections import OrderedDict
+
+        out = OrderedDict()
+        # all results are 1-d vectors
+        dims = np.array([1], dtype=np.intp)
+
+        nattr = len(attr_names)
+
+        from .main import PyQuery
+        from .subarray import Subarray
+
+        q = PyQuery(self._ctx_(), self, tuple(attr_names), tuple(), layout, False)
+        self.pyquery = q
+
+        if cond is not None and cond != "":
+            from .query_condition import QueryCondition
+
+            if isinstance(cond, str):
+                q.set_cond(QueryCondition(cond))
+            else:
+                raise TypeError("`cond` expects type str.")
+
+        if self.mode == "r":
+            q.set_subarray(subarray)
+
+        q.submit()
+
+        if self.mode == "d":
+            return
+
+        results = OrderedDict()
+        results = q.results()
+
+        # collect a list of dtypes for resulting to construct array
+        dtypes = list()
+        for i in range(nattr):
+            name, final_name = attr_names[i], attr_names[i]
+            if name == "__attr":
+                final_name = ""
+            if self.schema._needs_var_buffer(name):
+                if len(results[name][1]) > 0:  # note: len(offsets) > 0
+                    arr = q.unpack_buffer(name, results[name][0], results[name][1])
+                else:
+                    arr = results[name][0]
+                    arr.dtype = self.schema.attr_or_dim_dtype(name)
+                out[final_name] = arr
+            else:
+                if self.schema.domain.has_dim(name):
+                    el_dtype = self.schema.domain.dim(name).dtype
+                else:
+                    el_dtype = self.attr(name).dtype
+                arr = results[name][0]
+
+                # this is a work-around for NumPy restrictions removed in 1.16
+                if el_dtype == np.dtype("S0"):
+                    out[final_name] = b""
+                elif el_dtype == np.dtype("U0"):
+                    out[final_name] = ""
+                else:
+                    arr.dtype = el_dtype
+                    out[final_name] = arr
+
+            if self.schema.has_attr(final_name) and self.attr(final_name).isnullable:
+                out[final_name] = np.ma.array(
+                    out[final_name], mask=~results[name][2].astype(bool)
+                )
+
+        return out
+
+    def unique_dim_values(self, dim=None):
+        if dim is not None and not isinstance(dim, str):
+            raise ValueError("Given Dimension {} is not a string.".format(dim))
+
+        if dim is not None and not self.domain.has_dim(dim):
+            raise ValueError("Array does not contain Dimension '{}'.".format(dim))
+
+        query = self.query(attrs=[])[:]
+
+        if dim:
+            dim_values = tuple(np.unique(query[dim]))
+        else:
+            from collections import OrderedDict
```

Comment:
Do we want to move all these `from collections import OrderedDict`  imports to the top level?